Add Lifetime to enable observation of object deallocation in Swift

sharplet opened the pull request:
Because `rac_willDeallocSignal` doesn't have a native Swift equivalent yet, I've implemented it myself a few times in my projects. It's always felt a little complicated, and it only works for Objective-C objects.

I had an idea for a much simpler API, which works for Swift objects (and Objective-C ones), and is much simpler than `rac_willDeallocSignal`. I've seen other examples of this pattern in the wild — using `Signal.pipe()` to send a completed event on `deinit`. This PR is my take on that pattern, giving it the name `Lifetime`.

I've added a `lifetime` property to all `NSObject`s in a public extension that uses associated objects for storage.

I also added a new operator, `takeWithin`, that allows you to say the inverse of _take until this signal triggers_ — instead you can phrase it as _take within this object's lifetime_.

I'd love to know what you think!

andersio commented:
Just want to note that this is not the equivalent of `rac_willDeallocSignal`, since the token's deinitializer is called after its owner. If the observer uses `[unowned self]`, this might lead to surprising behaviour.

andersio commented:
Here is an example:

Say if I am observing an `NSManagedObjectContext`'s deallocation so as to remove an observer from the `NotificationCenter`, using this lifetime token would not do the job actually, since the weakly reference (or unowned) `self` needed by `removeObserver` by that time would be `nil`.

sharplet commented:
Yeah, UI bindings are what I had in mind.

Your example's a good one, I guess this implementation isn't a replacement for `rac_willDealloc`, as it's more of a _did_ dealloc notification.

mdiep commented:
I don't think the addition of the `Lifetime` type here is an improvement over the existing pattern. It's neither more nor less code, and it's less generic/flexible.

The lifetime semantics that @andersio mention are also a big concern.

How does this improve things for Swift objects?

NachoSoto commented on ReactiveCocoa/Swift/Lifetime.swift:
```diff
@@ -0,0 +1,36 @@
+
+	deinit {
+		endedObserver.sendCompleted()
+	}
```
This is redundant, but I guess it's good to be explicit here.

mdiep replied:
How is it redundant? I don't see how it would complete otherwise.

NachoSoto commented on ReactiveCocoa/Swift/Lifetime.swift:
```diff
@@ -0,0 +1,36 @@
+extension NSObject {
+	/// Returns a lifetime that ends when the receiver is deallocated.
+	@nonobjc public var lifetime: Lifetime {
+		if let lifetime = objc_getAssociatedObject(self, &lifetimeKey) as? Lifetime {
```
This should be `as! Lifetime?`.

sharplet replied:
Good catch 👍

NachoSoto commented on ReactiveCocoa/Swift/Signal.swift:
```diff
@@ -821,6 +821,13 @@ extension SignalType {
+	/// Forwards events from `self` until `lifetime` ends, at which point the
+	/// returned signal will complete.
+	@warn_unused_result(message="Did you forget to call `observe` on the signal?")
+	public func takeWithin(lifetime: Lifetime) -> Signal<Value, Error> {
```
Why not make this take an `NSObject` instead, and make `Lifetime` an implementation detail?

mdiep replied:
> Why not make this take an NSObject instead, and make Lifetime an implementation detail?

I think the idea is to make this usable with non-`NSObject` classes.

How about naming this `take(during lifetime: Lifetime)`. That seems like it would read a little more naturally.

```swift
.take(during: myObject.lifetime)
```

NachoSoto commented on ReactiveCocoa/Swift/Lifetime.swift:
```diff
@@ -0,0 +1,36 @@
+
+extension NSObject {
+	/// Returns a lifetime that ends when the receiver is deallocated.
+	@nonobjc public var lifetime: Lifetime {
```
TIL `@nonobjc`.

NachoSoto commented:
I like the idea of finally exposing this 👍 but I agree with @mdiep, I don't see why we need to expose `Lifetime`, it's an implementation detail.

I want to note that this is already implemented for [`SignalProducer.replayLazily`](https://github.com/ReactiveCocoa/ReactiveCocoa/blob/master/ReactiveCocoa/Swift/SignalProducer.swift#L1503-L1509), so I'd implement this in terms of that.

andersio commented:
@NachoSoto AFAIK we have neither means to extend `AnyObject`, nor storage in protocols with de deinitialisers. So the only way to get a pre- or a post-dealloc signal in pure Swift classes is defining your own.

andersio commented:
My second thought is that this has a vastly limited scope for only pure Swift controllers and non-Cocoa users (which we probably have none yet).

Under Cocoa, users can always grab a `willDeallocSignal`, and I assume we gonna make a Swift re-implementation of it before we are ready to split ObjC and Swift. Moreover, we usually recommend the `MutableProperty` + flattening approach for reuseable cells, whereas `MutableProperty` already has the semantics of `Lifetime`.

andersio commented:
That's said I think it might worth creating a new protocol to cover the token suggested by @sharplet, `NSObject`s and also voluntarily conforming Swift classes. The target at call site is unambiguous across the cases: `.take(withinLifetime: object)` and `.take(withinLifetime: object.token)`.

sharplet commented:
@andersio:

> My second thought is that this has a vastly limited scope for only pure Swift controllers and non-Cocoa users (which we probably don't have one yet).

@mdiep:

> How does this improve things for Swift objects?

The intent was that if you needed this for a Swift value, it would be as easy as adding a property:

```swift
class MyClass {
  private let lifetime = Lifetime()
  
  func start() {
    observeStuff.takeWithin(lifetime).start()
  }
}
```

@NachoSoto:

> Why not make this take an NSObject instead, and make Lifetime an implementation detail?

At best, I think this idea might be slightly more _learnable_ by being modelled as a real type. That is, if you were to read `takeWithin`'s signature or see the `Lifetime` class, you'd hopefully grasp its purpose easily.

But you're right, it also may not be beneficial enough to justify its existence. It's not super discoverable in its current form.

---

I've worked on a few teams with less experienced RAC users and it seems pretty common to do things like starting an unbounded timer or `NSNotificationCenter` observation.

Originally I had some ideas around actually incorporating the idea of a "lifetime" into the type of a producer, and then preventing it from being started until the lifetime has been explicitly specified (e.g., something like `takeWithin(lifetime)`, or `MutableProperty` which has the same semantics implicitly).

This PR is actually the result of some ideas I've been having around modelling signal lifetimes, but stripped back to be the simplest useful change I could think of.

If we agree that it's not enough of an improvement I'm happy to close this for now. Thanks for your input!

andersio commented:
@sharplet I'd like to redact my comment. After performing a few tests, `rac_willDeallocSignal` would not provide the semantics I said if `weak` or `unowned` references are used. Only `Unmanaged` (`unsafe unretained`) references would work.

andersio commented:
I have a snap idea though.

Say we have a protocol hierarchy of `protocol LifetimeProviding: LifetimeObservable`.

`LifetimeProviding` requires a lifetime object, while providing default implementations for `LifetimeObservable`. `LifetimeObservable` requires a EOL signal, and was implemented for `NSObject` and various `RAC` constructs.

NachoSoto commented on ReactiveCocoa/Swift/Lifetime.swift:
```diff
@@ -0,0 +1,36 @@
+	/// A signal that sends a Completed event when the lifetime ends.
+	public let ended: Signal<(), NoError>
+
+	private let endedObserver: Observer<(), NoError>
```
Can you use `Signal<(), NoError>.Observer`? That way it's clearer to see the parallel between the signal and its observer :)

sharplet replied:
Done.

NachoSoto commented:
I want to note that `SignalProducer.replayLazily` would have to be changed to be implemented using this new operator.

mdiep commented:
> The intent was that if you needed this for a Swift value, it would be as easy as adding a property:

That makes sense now. 👍🏻

mdiep commented:
I'm into this. 👍🏻

There are some conflicts and a few unaddressed items above. Would you be up for updating this?

_(Sorry for the delay on this!)_

andersio commented on ReactiveCocoa/Swift/Lifetime.swift:
```diff
@@ -0,0 +1,36 @@
+extension NSObject {
+	/// Returns a lifetime that ends when the receiver is deallocated.
+	@nonobjc public var lifetime: Lifetime {
+		if let lifetime = objc_getAssociatedObject(self, &lifetimeKey) as! Lifetime? {
```
Apparently we could have a race condition here. This would need to be guarded by the `objc_sync_*` pair or some kind of locks.

sharplet replied:
Yep, that's a good call. I've pushed a change which synchronises on `self` here.

> P.S. Why not simply write as? Lifetime?

I originally used that, see [this discussion](https://github.com/ReactiveCocoa/ReactiveCocoa/pull/3030#discussion_r69509093). There is no valid reason for the type of the associated object to be something other than `Lifetime`. If it is something else, then it should be caught as early as possible in development.

andersio commented:
`NSObject.lifetime` would need thread safety. Otherwise LGTM now.  🎉

-
Second thought on protocols:
> A protocol like `LifetimeProviding` might be less useful, if Lifetime is part of the public API. Say if we write a function `func bind(to: Something, during lifetime: Lifetime`, there is already an implicit constraint of objects that can provide a lifetime.

andersio commented:
I can now see `take(within:)` (or `take(during:)`?) that takes a `Lifetime` is optimal. :tada:

Constraining it with a protocol doesn't seem necessary to me, since we have cases that would use a `Lifetime` object directly. e.g. `replayLazily` as mentioned by @NachoSoto.

sharplet commented:
@andersio I like `take(during:)`! I've pushed an amended version with the operator renamed.

sharplet commented:
Ok, I think I've addressed all the feedback. Thanks so much!

I think what this could really use now is some proper API documentation, so this addition is more discoverable than our release notes or autocompletion. Should I take a stab at that as part of this PR, or can that be deferred until closer to the next release?

sharplet commented:
Oh, and one other thing: do you think it would be appropriate to add a Swift-specific deprecation warning to either/both of `rac_willDeallocSignal` and `rac_deallocDisposable`, encouraging Swift users to adopt `Lifetime` instead?

sharplet commented:
FWIW I think a Swift-specific deprecation should be achievable with a combination of `NS_REFINED_FOR_SWIFT` and a new Swift API using an appropriate availability attribute and calling through to the original.

andersio commented:
The Objective-C APIs would be moved to a separate module (ReactiveObjC?), and IIRC the ultimate goal is to completely drop the dependency of the Swift land on ReactiveObjC. So I don't see the need to mark it `@unavailable`.

andersio commented on ReactiveCocoa/Swift/SignalProducer.swift:
```diff
@@ -1942,10 +1942,10 @@ extension SignalProducerType {
 		// underlying producer. This is necessary because `struct`s don't have
 		// `deinit`.
-		let token = DeallocationToken()
+		let token = NSObject()
```
I think we should use a `Lifetime` here directly. `NSObject.lifetime` is available only on Darwin platforms.

sharplet replied:
Using `NSObject` here didn't bother me because of the use of `NSLock` above (although I suppose that will have a counterpart in swift-corelibs-foundation).

I agree though, using `Lifetime` directly makes sense!

sharplet replied:
Done.

andersio commented on ReactiveCocoa/Swift/Lifetime.swift:
```diff
@@ -0,0 +1,36 @@
+/// observe when that object goes out of scope.
+public final class Lifetime {
+	/// A signal that sends a Completed event when the lifetime ends.
+	public let ended: Signal<(), NoError>
```
[Bikeshedding] We should use a noun here, shouldn't we?

mdiep commented:
Should we use a prefix for `NSObject.lifetime`? Since it's a property, the type can't be used to disambiguate and it may conflict with existing methods/properties. cc #3049

andersio commented:
True... It should be prefixed before we ever figure out a new approach.

andersio commented:
An alternative would be built around a protocol and avoid using a property.

sharplet commented:
> The Objective-C APIs would be moved to a separate module (ReactiveObjC?). So I don't see the need to mark it unavailable.

The repo split won't happen until RAC 5, which means (unless I'm mistaken) that this is likely to be part of a 4.x release. In my own projects, I've wrapped `rac_signalForSelector`/`rac_deallocDisposable` in a number of places in order to expose the functionality to Swift in a type-safe way. So I think have a deprecation warning (as opposed to marking the API unavailable) could prove useful for any other users who have the same kind of wrappers and would benefit from being aware of the new feature. Do you agree?

andersio commented:
True for the split planned for RAC 5. But there _could_ be people using the ObjC API in Swift though, without caring about our shiny Swift offerings.

sharplet commented:
Personally, prefixing the name bothers me (but only a little), and I feel it hampers readability. Having said that, I totally understand the concern and agree that libraries should be very careful forcing extensions of types we don't on users, without any workaround (other than _rename your API_).

I actually feel fairly strongly that the naming is very much _the whole point_ of this change, which is what's giving me uncomfortable feels about it. 😕

I have to wonder whether there is actually any instance of `NSObject.lifetime` out there amongst RAC's users? Problem is the only way to find out would seem to be shipping a potentially breaking change and waiting to _maybe_ find out.

andersio commented:
I don't like it either. But in 4.x everything (aka the two Foundation extensions and the Rex bindings...) is prefixed. So we should be consistent, shouldn't we?

That's said there is sadly no way (AFAIK) we can completely safe from collision, except prefixing `rac` that is safe... ehm... conventionally.

mdiep commented:
Unfortunately, I think we need the prefix. 😞 Once that's done, I'm 👍🏻 to merge.

(Also, FWIW I don't think there will be any more 4.x releases. We'll merge `RAC5` to `master` once Travis updates to ß3.)

andersio commented:
Last Minute Question:

Having written:
```
object.lifetime.ended.observeCompleted(observer.sendCompleted)
```
makes me wonder if we should provide a convenience `observeCompleted` on `Lifetime`. Seems a bit awkward at the call site.

andersio commented on ReactiveCocoa/Swift/Lifetime.swift:
```diff
@@ -0,0 +1,39 @@
+		}
+
+		let lifetime = Lifetime()
+		objc_setAssociatedObject(self, &lifetimeKey, lifetime, .OBJC_ASSOCIATION_RETAIN)
```
I suggest using `RETAIN_NONATOMIC` instead. Using `RETAIN` (atomic) would somehow hinder the ability to use a `unowned(unsafe)` reference to the object.

mdiep commented:
@sharplet Are you interested in finishing this up? Not trying to pressure you; I just don't want this to fall between the cracks. 😄

sharplet commented:
Ok I think that's everything, except for this comment:

> [Bikeshedding] We should use a noun here, shouldn't we?

FWIW I really like the way `takeUntil(lifetime.ended)` reads. Although I'm certainly open to suggestions.

In any case, thanks for the nudge @mdiep and let me know if there's anything else! ✌️

mdiep commented:
This is great! Thanks @sharplet! 🎉

sharplet commented:
Thanks for merging @mdiep! And thanks everyone for the great feedback. I'd like to contribute some more documentation around this in the form of a usage guide, if you think that's appropriate. Perhaps an addition to the [Framework Overview](https://github.com/ReactiveCocoa/ReactiveCocoa/blob/master/Documentation/FrameworkOverview.md)?

mdiep commented:
Documentation is always appreciated. 😄
